daemon/containerd/image_commit.go

```diff
@@ -312,16 +319,43 @@ func applyDiffLayer(ctx context.Context, name string, baseImg imagespec.DockerOC
 		}
 	}()
 
-	if _, err = differ.Apply(ctx, diffDesc, mount); err != nil {
+	if _, err = differ.Apply(ctx, diffDesc, mounts); err != nil {
 		return err
 	}
 
-	if err = sn.Commit(ctx, name, key); err != nil {
-		if cerrdefs.IsAlreadyExists(err) {
-			return nil
+	if !i.idMapping.Empty() {
+		rootPair := i.idMapping.RootPair()
+		usernsID := fmt.Sprintf("%s-%d-%d", name, rootPair.UID, rootPair.GID)
+		remappedID := usernsID + remapSuffix
+
+		if err = sn.Commit(ctx, name+"-pre", key); err != nil {
+			if cerrdefs.IsAlreadyExists(err) {
+				return nil
+			}
+			return err
+		}
+
+		mounts, err = sn.Prepare(ctx, remappedID, name+"-pre")
+		if err != nil {
+			return err
+		}
+
+		if err := i.unremapRootFS(ctx, mounts); err != nil {
+			return err
+		}
+
+		if err := sn.Commit(ctx, name, remappedID); err != nil {
+			return err
+		}
+	} else {
+		if err = sn.Commit(ctx, name, key); err != nil {
+			if cerrdefs.IsAlreadyExists(err) {
+				return nil
+			}
+			return err
 		}
-		return err
 	}
```

Comment:
Would the `cerrdefs.IsAlreadyExists(err)` also be possible in the "remapped" case?


I'm considering if we should remove the `else`, and put the `sn.Commit` outside of the `if` (share it between both cases)

Reply:
Ha indeed, we can move the two commits outside, will do